samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -193,88 +215,133 @@ private void validateJoinQuery(LogicalJoin join, JoinInputNode.InputType inputTy
+        .map(x -> x - tableStartIndex) // re-adjust the offset
+        .sorted()
+        .collect(Collectors.toList()); // we have a list with all the input from table side with 0 based index.
+
+    // Validate the Condition must contain a ref to remote table primary key column.
+
+    if (conjunctionList.size() != 1 || tableRefsIdx.size() != 1) {
+      //TODO We can relax this by allowing another filter to be evaluated post lookup see SAMZA-2554
+      throw new SamzaException(
+          "Invalid query for join condition must contain exactly one predicate for remote table on __key__ column "
+              + dumpRelPlanForNode(join));
+    }
 
-    // Only reference operands are supported in row expressions and not constants.
-    // a.key = b.key is supported with a.key and b.key being reference operands.
-    // a.key = "constant" is not yet supported.
-    if (!(operands.get(0) instanceof RexInputRef) || !(operands.get(1) instanceof RexInputRef)) {
-      throw new SamzaException("SQL query is not supported. Join condition " + join.getCondition() + " should have "
-          + "reference operands but the types are " + operands.get(0).getClass() + " and " + operands.get(1).getClass());
+    // Validate the Project, follow each input and ensure that it is a simple ref with no rexCall in the way.
+    if (!isValidRemoteJoinRef(tableRefsIdx.get(0), isTablePosOnRight ? join.getRight() : join.getLeft())) {
+      throw new SamzaException("Invalid query for join condition can not have an expression and must be reference "
+          + SamzaSqlRelMessage.KEY_NAME + " column " + dumpRelPlanForNode(join));
     }
+  }
 
-    // Join condition is commutative, meaning, a.key = b.key is equivalent to b.key = a.key.
-    // Calcite assigns the indices to the fields based on the order a and b are specified in
-    // the sql 'from' clause. Let's put the operand with smaller index in leftRef and larger
-    // index in rightRef so that the order of operands in the join condition is in the order
-    // the stream and table are specified in the 'from' clause.
+  /**
+   * Helper method to check if the join condition can be evaluated by the remote table.
+   * It does follow single path  using the index ref path checking if it is a simple reference all the way to table scan.
+   * In case any RexCall is encountered will stop an return null as a marker otherwise will return Column Name.
+   *
+   * @param inputRexIndex rex ref index
+   * @param relNode current Rel Node
+   * @return false if any Relational Expression is encountered on the path, true if is simple ref to __key__ column.
+   */
+  private static boolean isValidRemoteJoinRef(int inputRexIndex, RelNode relNode){
+    if (relNode instanceof TableScan) {
+      return relNode.getRowType().getFieldList().get(inputRexIndex).getName().equals(SamzaSqlRelMessage.KEY_NAME);
+    }
+    // has to be a single rel kind filter/project/table scan
```

Comment:
could we add tests that go thru' the below code and fail ? If you plan to add them as part of your nested struct support, please feel free to drop this.

Reply:
@atoomula  this is hard to get done here because we can not handle any operator between TS and Join yet aka the NPE will come first, thus will add test for this case later.